src/v/config/constraints.cc

```diff
@@ -45,6 +54,534 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+              cluster_property_clamp(
+                topic_val,
+                cluster_opt,
+                enabled,
+                topic,
+                topic_property,
+                cluster_property);
+          },
+          [&topic_val, &topic, &topic_property](const auto range) {
+              range_clamp(topic_val, range, topic, topic_property);
+          });
+    }
+}
+
+template<typename T>
+void clamp_value(
+  std::optional<T>& topic_opt,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    auto tri = tristate<T>{topic_opt};
+    clamp_value(
+      tri, constraint, topic, topic_property, cluster_property, cluster_opt);
+    // The tristate is not in disabled state since it was assigned an optional
+    // earlier.
+    topic_opt = tri.get_optional();
+}
+} // namespace
+
+void constraint_clamp_topic_config(
+  cluster::topic_configuration& topic_cfg, const constraint_t& constraint) {
+    auto partition_count_tri = tristate(
+      std::make_optional(topic_cfg.partition_count));
+    auto replication_factor_tri = tristate(
+      std::make_optional(topic_cfg.replication_factor));
+    clamp_value(
+      partition_count_tri,
+      constraint,
+      topic_cfg.tp_ns.tp,
+      kafka::topic_property_partition_count,
+      config::shard_local_cfg().default_topic_partitions.name());
+    topic_cfg.partition_count = partition_count_tri.value();
+    clamp_value(
+      replication_factor_tri,
+      constraint,
+      topic_cfg.tp_ns.tp,
+      kafka::topic_property_replication_factor,
+      config::shard_local_cfg().default_topic_replication.name());
+    topic_cfg.replication_factor = replication_factor_tri.value();
+    clamp_value<model::compression>(
+      topic_cfg.properties.compression,
+      constraint,
+      topic_cfg.tp_ns.tp,
+      kafka::topic_property_compression,
+      config::shard_local_cfg().log_compression_type.name(),
+      config::shard_local_cfg().log_compression_type());
+    clamp_value<model::cleanup_policy_bitflags>(
+      topic_cfg.properties.cleanup_policy_bitflags,
+      constraint,
+      topic_cfg.tp_ns.tp,
+      kafka::topic_property_cleanup_policy,
+      config::shard_local_cfg().log_cleanup_policy.name(),
+      config::shard_local_cfg().log_cleanup_policy());
+    clamp_value<model::timestamp_type>(
+      topic_cfg.properties.timestamp_type,
+      constraint,
+      topic_cfg.tp_ns.tp,
+      kafka::topic_property_timestamp_type,
+      config::shard_local_cfg().log_message_timestamp_type.name(),
+      config::shard_local_cfg().log_message_timestamp_type());
+    clamp_value(
+      topic_cfg.properties.segment_size,
+      constraint,
+      topic_cfg.tp_ns.tp,
+      kafka::topic_property_segment_size,
+      config::shard_local_cfg().log_segment_size.name());
```

Comment:
It took me a little while, but I checked that these are all essentially the same for `constraint_clamp_topic_config` and `topic_config_satisfies_constraint`.

I wonder if it would be worth extracting these properties into traits classes or something similar, so that they can be reused.